File: src/com/l2kt/gameserver/handler/skillhandlers/BalanceLife.java
package com.l2kt.gameserver.handler.skillhandlers;

import java.util.ArrayList;
import java.util.List;

import com.l2kt.gameserver.handler.ISkillHandler;
import com.l2kt.gameserver.handler.SkillHandler;
import com.l2kt.gameserver.model.L2Skill;
import com.l2kt.gameserver.model.WorldObject;
import com.l2kt.gameserver.model.actor.Creature;
import com.l2kt.gameserver.model.actor.instance.Player;
import com.l2kt.gameserver.network.serverpackets.StatusUpdate;
import com.l2kt.gameserver.templates.skills.L2SkillType;

public class BalanceLife implements ISkillHandler
{
	private static final L2SkillType[] SKILL_IDS =
	{
		L2SkillType.BALANCE_LIFE
	};
	
	@Override
	public void useSkill(Creature activeChar, L2Skill skill, WorldObject[] targets)
	{
		final ISkillHandler handler = SkillHandler.getInstance().getHandler(L2SkillType.BUFF);
		if (handler != null)
			handler.useSkill(activeChar, skill, targets);
		
		final Player player = activeChar.getActingPlayer();
		final List<Creature> finalList = new ArrayList<>();
		
		double fullHP = 0;
		double currentHPs = 0;
		
		for (WorldObject obj : targets)
		{
			if (!(obj instanceof Creature))
				continue;
			
			final Creature target = ((Creature) obj);
			if (target.isDead())
				continue;
			
			// Player holding a cursed weapon can't be healed and can't heal
			if (target != activeChar)
			{
				if (target instanceof Player && ((Player) target).isCursedWeaponEquipped())
					continue;
				else if (player != null && player.isCursedWeaponEquipped())
					continue;
			}
			
			fullHP += target.getMaxHp();
			currentHPs += target.getCurrentHp();
			
			// Add the character to the final list.
			finalList.add(target);
		}
		
		if (!finalList.isEmpty())
		{
			double percentHP = currentHPs / fullHP;
			
			for (Creature target : finalList)
			{
				target.setCurrentHp(target.getMaxHp() * percentHP);
				
				StatusUpdate su = new StatusUpdate(target);
				su.addAttribute(StatusUpdate.CUR_HP, (int) target.getCurrentHp());
				target.sendPacket(su);
			}
		}
	}
	
	@Override
	public L2SkillType[] getSkillIds()
	{
		return SKILL_IDS;
	}
}

File: src/com/l2kt/gameserver/handler/skillhandlers/BalanceLife.kt
package com.l2kt.gameserver.handler.skillhandlers

import com.l2kt.gameserver.handler.ISkillHandler
import com.l2kt.gameserver.handler.SkillHandler
import com.l2kt.gameserver.model.L2Skill
import com.l2kt.gameserver.model.WorldObject
import com.l2kt.gameserver.model.actor.Creature
import com.l2kt.gameserver.model.actor.instance.Player
import com.l2kt.gameserver.network.serverpackets.StatusUpdate
import com.l2kt.gameserver.templates.skills.L2SkillType
import java.util.*

class BalanceLife : ISkillHandler {

    override val skillIds: Array<L2SkillType>
        get() = SKILL_IDS

    override fun useSkill(activeChar: Creature, skill: L2Skill, targets: Array<WorldObject>) {
        val handler = SkillHandler.getHandler(L2SkillType.BUFF)
        handler?.useSkill(activeChar, skill, targets)

        val player = activeChar.actingPlayer
        val finalList = ArrayList<Creature>()

        var fullHP = 0.0
        var currentHPs = 0.0

        for (obj in targets) {
            if (obj !is Creature)
                continue

            if (obj.isDead)
                continue

            // Player holding a cursed weapon can't be healed and can't heal
            if (obj !== activeChar) {
                if (obj is Player && obj.isCursedWeaponEquipped)
                    continue
                else if (player != null && player.isCursedWeaponEquipped)
                    continue
            }

            fullHP += obj.maxHp.toDouble()
            currentHPs += obj.currentHp

            // Add the character to the final list.
            finalList.add(obj)
        }

        if (!finalList.isEmpty()) {
            val percentHP = currentHPs / fullHP

            for (target in finalList) {
                target.currentHp = target.maxHp * percentHP

                val su = StatusUpdate(target)
                su.addAttribute(StatusUpdate.CUR_HP, target.currentHp.toInt())
                target.sendPacket(su)
            }
        }
    }

    companion object {
        private val SKILL_IDS = arrayOf(L2SkillType.BALANCE_LIFE)
    }
}